daffodil-runtime2/src/test/scala/org/apache/daffodil/runtime2/TestCodeGenerator.scala

```diff
@@ -168,4 +174,40 @@ class TestCodeGenerator {
     assert(ur.isError, "expected ur.isError to be true")
     assert(ur.getDiagnostics.nonEmpty, "expected ur.getDiagnostics to be non-empty")
   }
+
+  private def updateGeneratedCodeExample(schemaFile: os.Path, rootName: Option[String],
+                                         exampleCodeHeader: os.Path, exampleCodeFile: os.Path): Unit = {
+    // Generate code from the example schema file
+    val pf = Compiler().compileFile(schemaFile.toIO, rootName)
+    assert(!pf.isError, pf.getDiagnostics.map(_.getMessage()).mkString("\n"))
+    val cg = pf.forLanguage("c")
+    val rootNS = QName.refQNameFromExtendedSyntax(rootName.getOrElse("")).toOption
+    val codeDir = cg.generateCode(rootNS, tempDir.toString)
+    assert(!cg.isError, cg.getDiagnostics.map(_.getMessage()).mkString("\n"))
+    val generatedCodeHeader = codeDir/"libruntime"/"generated_code.h"
+    val generatedCodeFile = codeDir/"libruntime"/"generated_code.c"
+
+    // Replace the example generated code files
+    os.copy(generatedCodeHeader, exampleCodeHeader, replaceExisting = true)
+    os.copy(generatedCodeFile, exampleCodeFile, replaceExisting = true)
+  }
```

Comment:
We already have roundtrip parse/unparse tests for a number of schemas that should have sufficient coverage to let us know if something broke in the generated code.  I want to eliminate unnecessary steps to reduce friction for code contributors making changes to the C code generator.    I would rather have the tests update the expected files automatically instead of forcing code contributors to update the expected files manually.  Once the expected files are updated, code contributors will see any changes in `git diff` output and reviewers will see the changes in a pull request if the code contributor follows our code contributor workflow documentation (e.g., always run the tests first).  If a contributor changes some generator code but doesn't run the tests, the worst case is that those changes will appear in someone else's unrelated pull request, which still will alert us that the changes happened (they won't sneak past or avoid our review for too long).

Reply:
> the worst case is that those changes will appear in someone else's unrelated pull request,

This is what I hope to avoid. I actually find the example updates really helpful when doing code reviews--it's much easier to visualize what's changing compared to looking only at template C-code. Tests failing is a good way to know if the examples were updated to match the code changes and what we are reviewing is accurate.

If the goal is friction reduction, what if we add a new sbt command that runs and updates the examples, e.g. `sbt genExamples`. This sort of matches our existing `genProps` and `genSchemas` sbt commands, the only difference is we expect people to actually commit those files. And then the test to check this could just be a github action that's something like:

```
sbt genExamples && git diff --exit-code daffodil-runtime2/
``` 

Note that what you have is fine, and I won't block the PR over this. And if you do find this idea ok, it doesn't have to be part of this PR--sbt can be a beast sometimes so this probably won't be trivial to implement.

Reply:
I've made the test fail if the example generated code files were changed in any way.  The test still updates these files automatically to reduce friction.  All the code contributor has to do is to run "sbt test" before committing changes to a pull request as before, but now their own pull request will fail its tests instead of the changes appearing in someone else's unrelated pull request if they don't remember.

Reply:
So, if I sbt test, and the output has changed, the tests fail, but a side effect of running those tests, is that the comparison files are updated, so if I sbt test again, it passes?

This seems very problematic to me. 

Make 'sbt test' always idempotent. Same behavior every time. Updating output files to be match new stuff *should* be a manual step. You can provide some automation support for this to make it a one-shot action to update them all, but having 'sbt test' be non-idempotent is asking for trouble.

Reply:
Okay, then I want to remove this test and make "sbt compile" update the example code files instead.  We should automatically generate (a.k.a. update) the example code files whenever we build Daffodil, just like the build automatically generates those other files and keeps them up to date after any changes to the source code:

```scala
lazy val genManaged = taskKey[Unit]("Generate managed sources and resources")
lazy val genProps = taskKey[Seq[File]]("Generate properties scala source")
lazy val genSchemas = taskKey[Seq[File]]("Generated DFDL schemas")
```

All I want is to have pull requests include any changes to the example code files automatically.  I don't want any extra manual step.  We have parse and unparse tests which compile and run the new generated files.  These tests will warn us if any code generator backend changes break the generated files, so there is no reason why we need to compare the new generated files to the old files and fail the build even if only one byte differs.  It just would be nice (and helpful for reviewers) to show how code generator backend changes affect the example code files.

Steve (or Mike), what's the easiest way to make "sbt compile" call `updateGeneratedCodeExample` without needing to stage the daffodil CLI?